File: app/src/main/java/com/johnguant/redditthing/redditapi/ThingDeserializer.java
package com.johnguant.redditthing.redditapi;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.johnguant.redditthing.redditapi.model.Link;
import com.johnguant.redditthing.redditapi.model.Thing;

import java.lang.reflect.Type;

public class ThingDeserializer implements JsonDeserializer<Thing> {
    @Override
    public Thing deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonObject jObject = json.getAsJsonObject();
        switch (jObject.getAsJsonPrimitive("kind").getAsString()) {
            case "t3":
                Link link = context.deserialize(jObject.getAsJsonObject("data"), Link.class);
                return new Thing<>(jObject.getAsJsonPrimitive("kind").getAsString(), link);
        }
        return null;
    }
}

File: app/src/main/java/com/johnguant/redditthing/redditapi/ThingDeserializer.kt
package com.johnguant.redditthing.redditapi

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.johnguant.redditthing.redditapi.model.Link
import com.johnguant.redditthing.redditapi.model.Thing

import java.lang.reflect.Type

class ThingDeserializer : JsonDeserializer<Thing<*>> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Thing<*>? {
        val jObject = json.asJsonObject
        when (jObject.getAsJsonPrimitive("kind").asString) {
            "t3" -> {
                val link = context.deserialize<Link>(jObject.getAsJsonObject("data"), Link::class.java)
                return Thing(jObject.getAsJsonPrimitive("kind").asString, link)
            }
        }
        return null
    }
}
